File: src/auth/policy.ts
import type { AuthContext } from "./bearer-auth.ts";

export function scopesFromClaim(scope: unknown): string[] {
  if (typeof scope === "string") {
    return scope.split(/\s+/).filter(Boolean);
  }
  if (Array.isArray(scope)) {
    return scope.filter((entry): entry is string => typeof entry === "string");
  }
  return [];
}

export function groupsFromClaim(groups: unknown): string[] {
  if (!Array.isArray(groups)) {
    return [];
  }
  return groups.filter((entry): entry is string => typeof entry === "string");
}

export function hasScope(auth: Pick<AuthContext, "scopes">, scope: string): boolean {
  return auth.scopes.includes(scope);
}

export function hasGroup(auth: Pick<AuthContext, "groups">, group: string): boolean {
  return auth.groups.includes(group) || auth.groups.includes(`/${group}`);
}

export function canListTools(auth: Pick<AuthContext, "scopes">): boolean {
  return hasScope(auth, "mcp:tools:read");
}

export function canCallTools(auth: Pick<AuthContext, "scopes">): boolean {
  return hasScope(auth, "mcp:tools:execute");
}

export function canCallAdminTool(auth: Pick<AuthContext, "scopes" | "groups">): boolean {
  return hasScope(auth, "mcp:admin") || hasGroup(auth, "admin");
}

File: policy.go
package main

import "strings"

func scopesFromClaim(scope any) []string {
	switch typed := scope.(type) {
	case string:
		return strings.Fields(typed)
	case []any:
		values := make([]string, 0, len(typed))
		for _, entry := range typed {
			if value, ok := entry.(string); ok {
				values = append(values, value)
			}
		}
		return values
	default:
		return nil
	}
}

func groupsFromClaim(groups any) []string {
	entries, ok := groups.([]any)
	if !ok {
		return nil
	}
	values := make([]string, 0, len(entries))
	for _, entry := range entries {
		if value, ok := entry.(string); ok {
			values = append(values, value)
		}
	}
	return values
}

func hasScope(auth AuthContext, scope string) bool {
	for _, candidate := range auth.Scopes {
		if candidate == scope {
			return true
		}
	}
	return false
}

func hasGroup(auth AuthContext, group string) bool {
	for _, candidate := range auth.Groups {
		if candidate == group || candidate == "/"+group {
			return true
		}
	}
	return false
}

func canListTools(auth AuthContext) bool { return hasScope(auth, "mcp:tools:read") }

func canCallTools(auth AuthContext) bool { return hasScope(auth, "mcp:tools:execute") }

func canCallAdminTool(auth AuthContext) bool {
	return hasScope(auth, "mcp:admin") || hasGroup(auth, "admin")
}
